### ISSUE TITLE
Adding support for tunneling an existing TLS tunnel through tor

### ISSUE THREAD

**Issue by eayin2** [meejah/txtorcon#156]
I'd like to tunnel my TLS tunnel through tor, so that overall i'd have a tunnel like this:
tor(TLS(my_protocol)). Not sure if this is possible, maybe a convenience method (server and client side) to start TLS later is the way to add support for it.

Example of my current TLS endpoint (without tor tunnel) for the client:

`from __future__ import print_function
from twisted.internet.task import react
from twisted.internet.defer import inlineCallbacks
from twisted.internet.endpoints import TCP4ClientEndpoint

from twisted.python.filepath import FilePath
from twisted.internet import ssl, task, protocol, endpoints, defer
from twisted.python.modules import getModule

import my_protocol

@defer.inlineCallbacks
def main(reactor):
    factory = protocol.Factory.forProtocol(tmy_protocol.DecisionProtocol)
    ca_data = FilePath(b'ca_cert.pem').getContent()
    client_data = FilePath(b'phpc01.client.pem').getContent()
    ca_cert = ssl.Certificate.loadPEM(ca_data)
    client_key = ssl.PrivateCertificate.loadPEM(client_data)
    options = ssl.optionsForClientTLS(u'the-authority', ca_cert, client_key)
    endpoint = endpoints.SSL4ClientEndpoint(reactor, '46.101.200.90', 8966, options)
    echoClient = yield endpoint.connect(factory)
    done = defer.Deferred()
    echoClient.connectionLost = lambda reason: done.callback(None)
    yield done

task.react(main)
`

**Comment by meejah** [comment]
Did you try something like:
```
endpoint = txtorcon.TorClientEndpoint(server_ip, server_port)
tlsEndpoint = TLSWrapClientEndpoint(options, endpoint)
```

**Comment by meejah** [comment]
p.s. this would make a good addition to the ``examples/`` if you get it working :)

**Comment by eayin2** [comment]
Yes, that worked fine with txtorcon.TorClientEndpoint. Thanks for the hint. I had to use specifically provide socks_hostname='127.0.0.1', else it'd pick automatically the ipv6 localhost ::1 (i think).

# client.py
```
from __future__ import print_function

from twisted.internet.task import react
from twisted.internet.defer import inlineCallbacks
from twisted.internet.endpoints import TCP4ClientEndpoint
from twisted.python.filepath import FilePath
from twisted.internet import ssl, task, protocol, endpoints, defer
from twisted.python.modules import getModule
from txsocksx.tls import TLSWrapClientEndpoint
from txsocksx.client import SOCKS5ClientEndpoint
import txtorcon
import other_proto


@defer.inlineCallbacks
def main(reactor):
    factory = protocol.Factory.forProtocol(other_proto.DecisionProtocol)
    ca_data = FilePath(b'ca_cert.pem').getContent()
    client_data = FilePath(b'a.client.pem').getContent()
    ca_cert = ssl.Certificate.loadPEM(ca_data)
    client_key = ssl.PrivateCertificate.loadPEM(client_data)
    options = ssl.optionsForClientTLS(u'the-authority', ca_cert, client_key)
    exampleEndpoint = txtorcon.TorClientEndpoint(ip, 8966, socks_hostname="127.0.0.1")
    tlsEndpoint = TLSWrapClientEndpoint(options, exampleEndpoint)
    deferred = yield tlsEndpoint.connect(factory)
    done = defer.Deferred()
    deferred.connectionLost = lambda reason: done.callback(None)
    yield done

task.react(main)  # (1)
```

**Comment by xplsek03** [comment]
@meejah Do I understand correctly that in current version this behaviour could be achieved by using stream_via method, with argument socks_endpoint=SSL4ClientEndpointClass?

**Comment by xplsek03** [comment]
@meejah is it possible to achieve this behaviour using this on client?

`tor = yield txtorcon.launch()
#launch fresh tor
ep=tor.stream_via(host,port,tls=True,socks_endpoint=SSL4ClientEndpointTwistedClass)
#connect to server (host, port) using tls and already created twisted TLS instance
ep.connect(someFactory)`

From what I understand, using txsocksx is not necessary from 2017, but I couldn't find any other sources for this tls usage.
